spec/features/two_factor_authentication/sign_in_via_recovery_code_spec.rb

```diff
@@ -0,0 +1,38 @@
+require 'rails_helper'
+
+feature 'Signing in via recovery code' do
+  it 'displays new recovery code and redirects to profile after acknowledging' do
+    user = create(:user, :signed_up)
+    sign_in_before_2fa(user)
+
+    code = RecoveryCodeGenerator.new(user).create
+
+    click_link t('devise.two_factor_authentication.recovery_code_fallback.link')
+
+    enter_recovery_code(code: code)
+
+    click_submit_default
+    click_acknowledge_recovery_code
+
+    expect(user.reload.recovery_code).to_not eq code
+    expect(current_path).to eq profile_path
+  end
+
+  context 'user enters incorrect recovery code' do
+    it 'locks user out when max login attempts has been reached' do
```

Comment:
this is the new test :) the rest was me just re-arranging stuff because it was very difficult to tell which parts of recovery code were already tested given the organization.